api/errors/codes/codes.go

```diff
@@ -0,0 +1,331 @@
+	// by the underlying system have been broken. This error code is reserved
+	// for serious errors.
+	//
+	// HTTP Mapping: 500 Internal Server Error
+	// Google API Mapping: Code.INTERNAL
+	Internal Code = 13
+
+	// Unavailable means the service is currently unavailable. This is most likely a
+	// transient condition, which can be corrected by retrying with
+	// a backoff.
+	//
+	// See the guidelines above for deciding between `FailedPrecondition`,
+	// `Aborted`, and `Unavailable`.
+	//
+	// HTTP Mapping: 503 Service Unavailable
+	// Google API Mapping: Code.UNAVAILABLE
+	Unavailable Code = 14
+
+	// DataLoss means unrecoverable data loss or corruption.
+	//
+	// HTTP Mapping: 500 Internal Server Error
+	// Google API Mapping: Code.DATA_LOSS
+	DataLoss Code = 15
+
+	// Unauthenticated means the request does not have valid authentication
+	// credentials for the operation.
+	//
+	// HTTP Mapping: 401 Unauthorized
+	// Google API Mapping: Code.UNAUTHENTICATED
+	Unauthenticated Code = 16
+
+	// Application means there was an application error. This will typically be
+	// accompianed by a user-defined error name across the wire.
+	//
+	// HTTP Mapping: 500 Internal Server Error
+	// Google API Mapping: Code.UNKNOWN
+	Application Code = 17
+)
+
+var (
+	codeToString = map[Code]string{
+		None:               "none",
+		Cancelled:          "cancelled",
+		Unknown:            "unknown",
+		InvalidArgument:    "invalid-argument",
+		DeadlineExceeded:   "deadline-exceeded",
+		NotFound:           "not-found",
+		AlreadyExists:      "already-exists",
+		PermissionDenied:   "permission-denied",
+		ResourceExhausted:  "resource-exhausted",
+		FailedPrecondition: "failed-precondition",
+		Aborted:            "aborted",
+		OutOfRange:         "out-of-range",
+		Unimplemented:      "unimplemented",
+		Internal:           "internal",
+		Unavailable:        "unavailable",
+		DataLoss:           "data-loss",
+		Unauthenticated:    "unauthenticated",
+		Application:        "application",
+	}
+	stringToCode = map[string]Code{
+		"none":                None,
+		"cancelled":           Cancelled,
+		"unknown":             Unknown,
+		"invalid-argument":    InvalidArgument,
+		"deadline-exceeded":   DeadlineExceeded,
+		"not-found":           NotFound,
+		"already-exists":      AlreadyExists,
+		"permission-denied":   PermissionDenied,
+		"resource-exhausted":  ResourceExhausted,
+		"failed-precondition": FailedPrecondition,
+		"aborted":             Aborted,
+		"out-of-range":        OutOfRange,
+		"unimplemented":       Unimplemented,
+		"internal":            Internal,
+		"unavailable":         Unavailable,
+		"data-loss":           DataLoss,
+		"unauthenticated":     Unauthenticated,
+		"application":         Application,
+	}
+	codeToHTTPStatusCode = map[Code]int{
+		None:               200,
+		Cancelled:          499,
+		Unknown:            500,
+		InvalidArgument:    400,
+		DeadlineExceeded:   504,
+		NotFound:           404,
+		AlreadyExists:      409,
```

Comment:
Ya we can move this I guess, I figured it made some sense since a lot of people think of HTTP codes when they discuss errors.